metadata/component.go

```diff
@@ -94,3 +94,87 @@ func (m *manager) CreateComponent(name string, text string, params param.Params,
 	log.Debugf("Writing component parameters at '%s/%s", componentsDir, name)
 	return m.writeComponentParams(name, params)
 }
+
+func (m *manager) DeleteComponent(name string) error {
+	componentPath, err := m.findComponentPath(name)
+	if err != nil {
+		return err
+	}
+
+	// Delete the component file in components/
+	log.Infof("Deleting component '%s' at path '%s'", name, componentPath)
+	m.appFS.Remove(componentPath)
+
+	// Delete the component references.
+	log.Infof("Removing component parameter references ...")
+
+	// Part of removing the component references imply removing the references
+	// in component/params.libsonnet.
+	log.Debugf("... deleting references in %s", m.componentParamsPath)
+	componentParamsFile, err := afero.ReadFile(m.appFS, string(m.componentParamsPath))
+	if err != nil {
+		return err
+	}
+	jsonnet, err := param.DeleteComponent(name, string(componentParamsFile))
+	if err != nil {
+		return err
+	}
+	err = afero.WriteFile(m.appFS, string(m.componentParamsPath), []byte(jsonnet), defaultFilePermissions)
```

Comment:
I think we should move all FS operations to the very end of the function. Consider a scenario like: we succeed deleting this component, but then something goes wrong as we're parsing the env files, which is not unlikely given the limitations of our transformer at this point. So we abort. This leaves the component half-deleted, and their app in a generally unusable state.

So, overall I'd rather (1) perform all transformations to data, then (2) write out the files and remove things from the FS.

Thoughts?

Reply:
This makes sense.